platform/security/rest/security-rest-cxfwrapper/src/main/java/org/codice/ddf/cxf/oauth/OAuthSecurityImpl.java

```diff
@@ -0,0 +1,256 @@
+   * @param discoveryUrl the metadata URL of the OAuth provider
+   */
+  public void setUserTokenOnClient(
+      Client client,
+      Subject subject,
+      String sourceId,
+      String clientId,
+      String clientSecret,
+      String discoveryUrl) {
+    if (client == null || subject == null || Strings.isBlank(sourceId)) {
+      return;
+    }
+
+    String userId = SubjectUtils.getEmailAddress(subject);
+    if (userId == null) {
+      userId = SubjectUtils.getName(subject);
+    }
+
+    TokenInformation.TokenEntry tokenEntry = tokenStorage.read(userId, sourceId);
+    if (tokenEntry == null) {
+      return;
+    }
+
+    LOGGER.debug("Adding access token to OAUTH header.");
+    client.header(OAUTH, BEARER + tokenEntry.getAccessToken());
+  }
+
+  /**
+   * Gets the system's access token from the token storage to set it to the OAUTH header. If one can
+   * not be found, retrieves the system's access token from the configured OAuth provider. header
+   *
+   * @param client Non-null client to set the access token on.
+   * @param clientId The client ID registered with the OAuth provider
+   * @param clientSecret The client secret registered with the OAuth provider
+   * @param discoveryUrl the metadata URL of the OAuth provider
+   */
+  public void setSystemTokenOnClient(
+      Client client, String clientId, String clientSecret, String discoveryUrl) {
+    if (client == null
+        || Strings.isBlank(clientId)
+        || Strings.isBlank(clientSecret)
+        || Strings.isBlank(discoveryUrl)) {
+      return;
+    }
+
+    String encodedClientIdSecret =
+        Base64.getEncoder().encodeToString((clientId + ":" + clientSecret).getBytes(UTF_8));
+
+    TokenInformation.TokenEntry tokenEntry =
+        tokenStorage.read(encodedClientIdSecret, CLIENT_CREDENTIALS);
+
+    if (tokenEntry == null || !discoveryUrl.equalsIgnoreCase(tokenEntry.getDiscoveryUrl())) {
+      String accessToken = getNewAccessToken(encodedClientIdSecret, discoveryUrl);
+      if (accessToken == null) {
+        return;
+      }
+
+      client.header(OAUTH, BEARER + accessToken);
+      return;
+    }
+
+    if (isExpired(tokenEntry.getAccessToken())) {
+      String accessToken = getNewAccessToken(encodedClientIdSecret, discoveryUrl);
+      if (accessToken == null) {
+        return;
+      }
+
+      client.header(OAUTH, BEARER + accessToken);
+      return;
+    }
+
+    String accessToken = tokenEntry.getAccessToken();
+    client.header(OAUTH, BEARER + accessToken);
+  }
+
+  /**
+   * Gets an access token from the configured OAuth provider, saves it to the token storage and
+   * returns it
+   *
+   * @param encodedClientIdSecret - the base 64 encoded clientId:secret
+   * @param discoveryUrl - the URL where the Oauth provider's metadata is hosted
+   * @return a client access token or null if one could not be returned
+   */
+  private String getNewAccessToken(String encodedClientIdSecret, String discoveryUrl) {
```

Comment:
The refresh logic is not the same. In this case, we just get a new access token by passing in the client id and the secret cause we don't need the user to do that. In the OAuth plugin because we're using the user's token we actually do a refresh. The only thing that is similar is the `isExpired` method and I don't think it's worth creating a utils class just for that.